lib/logger/lib/logger/legacy_handler.ex

```diff
@@ -0,0 +1,265 @@
+defmodule Logger.LegacyHandler do
+  @moduledoc false
+
+  alias Logger.Counter
+
+  @internal_keys [:counter]
+
+  @doc false
+  def child_spec(start_options) do
+    %{
+      id: :gen_event,
+      start: {:gen_event, :start_link, [{:local, Logger}, start_options]},
+      modules: :dynamic
+    }
+  end
+
+  def adding_handler(%{config: data} = config) do
+    # TODO: When using counters exclusively then this line can be changed to use
+    # `Map.put/2` instead as we will not need to pass ETS table
+    new_data =
+      Logger.Config.default_config()
+      |> Map.merge(data || %{})
+      |> Map.put_new_lazy(:counter, &Logger.Config.new/0)
+
+    {:ok, %{config | config: new_data}}
+  end
+
+  # TODO: Remove when we will support OTP 22+
+  def changing_config(current, new), do: changing_config(:update, current, new)
+
+  def changing_config(
+        op,
+        %{id: id, module: module, config: old_data, filters: filters},
+        %{config: new_data} = new_config
+      ) do
+    {counter, old_data} = Map.pop(old_data, :counter)
+
+    old_data =
+      case op do
+        :set -> Logger.Config.default_config()
+        :update -> old_data
+      end
+
+    data =
+      old_data
+      |> Map.merge(new_data)
+      |> Map.put(:counter, counter)
+
+    config =
+      %{
+        id: id,
+        module: module,
+        level: :all,
+        filter_default: :log,
+        formatter: {:logger_formatter, %{}},
+        filters: filters
+      }
+      |> Map.merge(new_config)
+
+    {:ok, Map.put(config, :config, data)}
+  end
+
+  def filter_config(%{config: data} = config) do
+    %{config | config: Map.drop(data, @internal_keys)}
+  end
+
+  @doc """
+  Hook required by `:logger`.
+  """
+  def log(%{meta: %{domain: [:otp, :sasl | _]}}, %{config: %{sasl: false}}) do
+    :ok
+  end
+
+  def log(%{meta: %{domain: [:supervisor_report | _]}}, %{config: %{sasl: false}}) do
+    :ok
+  end
+
+  def log(%{level: erl_level, msg: msg, meta: erl_meta}, %{id: _id, config: config}) do
```

Comment:
We should have a clause that matches on the Elixir domain. If we are in the Elixir domain, we can skip many steps below:

1. We can skip `erlang_level_to_elixir_level`
2. We can skip the `min_level` check
3. We can skip the splitting of the Erlang metadata

There are probably others.

Reply:
Translations will probably be moved one layer upward to the formatters as IMHO that makes more sense and allow user to have better integration with other backends. For example imagine that user want to use `logger_disk_log_h` but want to have Elixir-like format used in that log - right now that will require user to rewrite whole formatting on their own, with formatter as a separate layer they could use just that module as formatter.

Reply:
That's a good point but I think we should go ahead and at least split them now. Otherwise performance will suffer and I want to avoid regression in behaviour (improvements later are always welcome though!).